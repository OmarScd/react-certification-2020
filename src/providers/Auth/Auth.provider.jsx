import React, { useState, useEffect, useContext } from 'react';

import { USERNAME_STORAGE_KEY } from '../../utils/constants';
import { storage } from '../../utils/storage';

const AuthContext = React.createContext(null);

const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error(`Can't use "useAuth" without an AuthProvider!`);
  }
  return context;
};

const AuthProvider = ({ children }) => {
  const [authenticated, setAuthenticated] = useState(false);
  const [username, setUsername] = useState(undefined);

  useEffect(() => {
    const user = storage.get(USERNAME_STORAGE_KEY);
    const isAuthenticated = !!user;

    setAuthenticated(isAuthenticated);
    if (isAuthenticated) {
      setUsername(user);
    }
  }, []);

  const login = (name) => {
    setAuthenticated(true);
    setUsername(name);
    storage.set(USERNAME_STORAGE_KEY, name);
  };

  const logout = () => {
    setAuthenticated(false);
    setUsername(undefined);
    storage.remove(USERNAME_STORAGE_KEY);
  };

  return (
    <AuthContext.Provider value={{ login, logout, authenticated, username }}>
      {children}
    </AuthContext.Provider>
  );
};

export { useAuth, AuthProvider as default };
