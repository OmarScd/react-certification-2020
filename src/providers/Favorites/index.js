export { default } from './Favorites.provider';
