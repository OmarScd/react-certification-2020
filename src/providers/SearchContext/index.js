export { default, useSearch } from './SearchContext.provider';
