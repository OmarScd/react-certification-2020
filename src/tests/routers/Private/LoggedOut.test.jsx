import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Private from '../../../routers/Private';
import '../../../providers/Auth';

jest.mock('../../../providers/Auth', () => ({
  default: jest.fn(),
  useAuth: () => ({ authenticated: false }),
}));

describe('Private', () => {
  afterAll(cleanup);

  it('renders redirect when not authenticated', () => {
    const { container } = render(
      <BrowserRouter>
        <Private />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });
});
