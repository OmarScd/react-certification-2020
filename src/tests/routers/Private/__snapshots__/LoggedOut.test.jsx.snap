// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Private renders redirect when not authenticated 1`] = `<div />`;
